### PR TITLE
Run phpunit tests

### LINKED ISSUE #11: Run phpunit locally
There are currently quite a few phpunit tests running around the Moodle codebase. They require bringing the kitchen sink with you to run - but they do test the functionality of the codebase and so I do think they are going to be valuable at least during the initial phases.

We should get the tests running in the docker container (hopefully with one command) so that they can easily be run by anybody working on the project.

### PR DESCRIPTION
**Pull request by JHeimbach**
Closes #11 

Command for legacy phpunit tests: `make legacyUnit`
Command for behat tests: `make behat`
Command for phpunit tests: `make unit`

I will add/edit the Documentation/How-To to the wiki.

### CODE REVIEW

**Comment by CarwynNelson** [comment]
Hav

File: lib/filestorage/file_exceptions.php [hunk 1]
```diff
@@ -24,6 +24,7 @@
 
 defined('MOODLE_INTERNAL') || die();
 
+require_once __DIR__ . '/../setuplib.php';
```
Comment (CarwynNelson) [inline lib/filestorage/file_exceptions.php]: What what the reason for needing this?

**Comment by CarwynNelson** [comment]
Please could you merge master into this branch? The `phpunit.xml.dist` file is out of date.

**Comment by CarwynNelson** [comment]
Looks like the test `bootstrap.php` is not set to include the `config.dev.php` file, which causes the following error to occur:

```
Fatal error: $CFG->dataroot is not configured properly, directory does not exist or is not accessible! Exiting.
make: *** [legacyPhpunitInstall] Error 1
```

**Comment by JHeimbach** [comment]
@CarwynNelson hey, sorry for your troubles, i think all you problems go away if you copy .env.dist to .env

I rewrote the wiki entry for local development to account for the changes I made. If something isn't clear enough, please let me know.

https://github.com/BetterMood/better-mood/wiki/Local-Development

File: lib/filestorage/file_exceptions.php (hunk 1 above)
Comment (CarwynNelson) [reply lib/filestorage/file_exceptions.php]: Sounds good to me.

**Comment by JHeimbach** [comment]
@CarwynNelson checked the repo in a new folder out and set it up as described in the wiki. everything seems working so far. i didn't let all tests run because that is a night job, but the configuration seems to be working as intended.

If you already have a docker system running, please delete it (`docker-compose rm -s`) and clear the volumes. (`docker volume prune` -> deletes all volumes that are not attached to a container).

Also try and pull the newest docker images (`docker pull bettermood/php`) or clone the repository bettermood/docker and build the images yourself (you can tag them as well).

The underlying problem is, that I use docker volumes so I don't have to install every time a new docker image is made (because it costs time), and owner of docker volumes are always root, if the folder is not exits before. in betttermood/php, I create the data folders and change the owner to www-data, the user that runs the php code.

**Comment by CarwynNelson** [comment]
So I've run `docker system prune a` (deletes everything basically) and re-ran `make legacyUnit`

Here is the output from running that command (downloading of docker images removed):
```
Creating better-mood_php_1      ... done
Creating better-mood_phpunit_1   ... done
Creating better-mood_exttests_1  ... done
Creating better-mood_mysql_1    ... done
Creating better-mood_selenium_1 ... done
Creating better-mood_mailhog_1   ... done
Creating better-mood_webserver_1 ... done
Creating better-mood_behat_1     ... done
sleep 10
docker-compose exec phpunit php admin/tool/phpunit/cli/init.php
Initialising Moodle PHPUnit test environment...
You are already using composer version 1.6.5 (stable channel).
Do not run Composer as root/super user! See https://getcomposer.org/root for details
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Nothing to install or update
Generating autoload files
Fatal error: $CFG->dataroot is not configured properly, directory does not exist or is not accessible! Exiting.
make: *** [legacyPhpunitInstall] Error 1
```

The important bit being this:
```
Fatal error: $CFG->dataroot is not configured properly, directory does not exist or is not accessible! Exiting.
```

I'm quite certain that the reason for this is that `config.dev.php` is not the active config. Is the docker container or make command meant to deal with that?

I'm confused as to how it's working for you if this config is never loaded anywhere explicitly.

**Comment by JHeimbach** [comment]
@CarwynNelson sorry forgot to mention it, you have to copy config.dev.php to config.php

**Comment by CarwynNelson** [comment]
Fantastic. The legacy tests are now running :)

Personally I'm happy for this to be merged. If any of the tests are failing then we can raise that as a new issue.

It would be best from a CI purity perspective to get these tests running on the travis pipeline (that's an issue for another ticket though :P). It would increase the build time but it would also reduce the likelihood of us breaking existing functionality.

**Comment by JHeimbach** [comment]
Has somebody some objection against this pr?

**Pull request merged**